File: src/main/java/com/vlog/vlog/post/PostService.java
package com.vlog.vlog.post;

import com.vlog.vlog.common.NotFoundException;
import com.vlog.vlog.post.dto.CreatePostRequest;
import com.vlog.vlog.post.dto.PostFilter;
import com.vlog.vlog.post.dto.PostResponse;
import com.vlog.vlog.user.User;
import com.vlog.vlog.user.UserService;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class PostService {

    private final PostRepository repository;
    private final UserService userService;

    public PostService(PostRepository repository, UserService userService) {
        this.repository = repository;
        this.userService = userService;
    }

    public Page<PostResponse> search(PostFilter filter, Pageable pageable) {
        Specification<Post> spec = PostSpecifications.build(filter);
        return repository.findAll(spec, pageable).map(PostResponse::from);
    }

    public PostResponse findById(Long id) {
        return repository.findById(id)
                .map(PostResponse::from)
                .orElseThrow(() -> new NotFoundException("Post not found: " + id));
    }

    @Transactional
    public PostResponse create(CreatePostRequest request, String authorUsername) {
        User author = userService.getEntityByUsername(authorUsername);

        Post post = new Post(
                request.title(),
                request.body(),
                author,
                request.postDate()
        );
        return PostResponse.from(repository.save(post));
    }

    @Transactional
    public PostResponse update(Long id, CreatePostRequest request, String username) {
        Post post = repository.findById(id)
                .orElseThrow(() -> new NotFoundException("Post not found: " + id));

        requireOwnerOrManager(post, username);

        post.update(request.title(), request.body(), request.postDate());
        return PostResponse.from(post);
    }

    @Transactional
    public void delete(Long id, String username) {
        Post post = repository.findById(id)
                .orElseThrow(() -> new NotFoundException("Post not found: " + id));

        requireOwnerOrManager(post, username);
        repository.delete(post);
    }

    public long countByAuthor(String username) {
        return repository.countByAuthorUsername(username);
    }

    /** Authors may modify their own vlogs; managers may modify any. */
    private void requireOwnerOrManager(Post post, String username) {
        if (post.getAuthor().getUsername().equals(username)) {
            return;
        }

        User caller = userService.getEntityByUsername(username);
        if (!caller.isManager()) {
            throw new AccessDeniedException("You can only modify your own vlogs");
        }
    }
}

File: vlog-frontend/src/app/post/post.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { API_BASE } from '../core/api.config';
import { CreatePostRequest, Page, PostFilter, PostResponse } from './post.model';

export interface PageRequest {
  page?: number;
  size?: number;
  sort?: string;          // e.g. 'postDate,desc'
}

@Injectable({ providedIn: 'root' })
export class PostService {

  private readonly http = inject(HttpClient);
  private readonly url = `${API_BASE}/posts`;

  search(filter: PostFilter = {}, page: PageRequest = {}): Observable<Page<PostResponse>> {
    let params = new HttpParams();

    for (const [key, value] of Object.entries({ ...filter, ...page })) {
      if (value !== null && value !== undefined && value !== '') {
        params = params.set(key, String(value));
      }
    }

    return this.http.get<Page<PostResponse>>(this.url, { params });
  }

  findById(id: number): Observable<PostResponse> {
    return this.http.get<PostResponse>(`${this.url}/${id}`);
  }

  create(request: CreatePostRequest): Observable<PostResponse> {
    return this.http.post<PostResponse>(this.url, request);
  }

  update(id: number, request: CreatePostRequest): Observable<PostResponse> {
    return this.http.put<PostResponse>(`${this.url}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.url}/${id}`);
  }

  countByAuthor(username: string): Observable<number> {
    return this.http.get<number>(`${this.url}/count`, {
      params: new HttpParams().set('username', username)
    });
  }
}
